# Day 18: Snailfish

## Part one

You descend into the ocean trench and encounter some [snailfish](https://en.wikipedia.org/wiki/Snailfish). They say they saw the sleigh keys! They'll even tell you which direction the keys went if you help one of the smaller snailfish with his **math homework**.

Snailfish numbers aren't like regular numbers. Instead, every snailfish number is a pair - an ordered list of two elements. Each element of the **pair** can be either a regular number or another pair.

Pairs are written as `[x,y]`, where x and y are the elements within the pair. Here are some example snailfish numbers, one snailfish number per line:

```
[1,2]
[[1,2],3]
[9,[8,7]]
[[1,9],[8,5]]
[[[[1,2],[3,4]],[[5,6],[7,8]]],9]
[[[9,[3,8]],[[0,9],6]],[[[3,7],[4,9]],3]]
[[[[1,3],[5,3]],[[1,3],[8,7]]],[[[4,9],[6,9]],[[8,2],[7,3]]]]
```

This snailfish homework is about **addition**. To add two snailfish numbers, form a pair from the left and right parameters of the addition operator. For example, `[1,2]` + `[[3,4],5]` becomes `[[1,2],[[3,4],5]]`.

There's only one problem: **snailfish numbers must always be reduced**, and the process of adding two snailfish numbers can result in snailfish numbers that need to be reduced.

To **reduce a snailfish number**, you must repeatedly do the first action in this list that applies to the snailfish number:

- If any pair is **nested inside four pairs**, the leftmost such pair **explodes**.
- If any regular number is **10 or greater**, the leftmost such regular number **splits**.

Once no action in the above list applies, the snailfish number is reduced.

During reduction, at most one action applies, after which the process returns to the top of the list of actions. For example, if **split** produces a pair that meets the **explode** criteria, that pair **explodes** before other **splits** occur.

To **explode** a pair, the pair's left value is added to the first regular number to the left of the exploding pair (if any), and the pair's right value is added to the first regular number to the right of the exploding pair (if any). Exploding pairs will always consist of two regular numbers. Then, the entire exploding pair is replaced with the regular number 0.

Here are some examples of a single explode action:

- `[[[[[9,8],1],2],3],4]` becomes `[[[[0,9],2],3],4]` (the 9 has no regular number to its left, so it is not added to any regular number).
- `[7,[6,[5,[4,[3,2]]]]]` becomes `[7,[6,[5,[7,0]]]]` (the 2 has no regular number to its right, and so it is not added to any regular number).
- `[[6,[5,[4,[3,2]]]],1]` becomes `[[6,[5,[7,0]]],3]`.
- `[[3,[2,[1,[7,3]]]],[6,[5,[4,[3,2]]]]]` becomes `[[3,[2,[8,0]]],[9,[5,[4,[3,2]]]]]` (the pair `[3,2]` is unaffected because the pair `[7,3]` is further to the left; `[3,2]` would explode on the next action).
- `[[3,[2,[8,0]]],[9,[5,[4,[3,2]]]]]` becomes `[[3,[2,[8,0]]],[9,[5,[7,0]]]]`.

To **split** a regular number, replace it with a pair; the left element of the pair should be the regular number divided by two and rounded **down**, while the right element of the pair should be the regular number divided by two and rounded **up**. For example, 10 becomes `[5,5]`, 11 becomes `[5,6]`, 12 becomes `[6,6]`, and so on.

Here is the process of finding the reduced result of `[[[[4,3],4],4],[7,[[8,4],9]]] + [1,1]`:

```
after addition: [[[[[4,3],4],4],[7,[[8,4],9]]],[1,1]]
after explode:  [[[[0,7],4],[7,[[8,4],9]]],[1,1]]
after explode:  [[[[0,7],4],[15,[0,13]]],[1,1]]
after split:    [[[[0,7],4],[[7,8],[0,13]]],[1,1]]
after split:    [[[[0,7],4],[[7,8],[0,[6,7]]]],[1,1]]
after explode:  [[[[0,7],4],[[7,8],[6,0]]],[8,1]]
```

Once no reduce actions apply, the snailfish number that remains is the actual result of the addition operation: `[[[[0,7],4],[[7,8],[6,0]]],[8,1]]`.

The homework assignment involves adding up a **list of snailfish numbers** (your puzzle input). The snailfish numbers are each listed on a separate line. Add the first snailfish number and the second, then add that result and the third, then add that result and the fourth, and so on until all numbers in the list have been used once.

For example, the final sum of this list is `[[[[1,1],[2,2]],[3,3]],[4,4]]`:

```
[1,1]
[2,2]
[3,3]
[4,4]
```

The final sum of this list is `[[[[3,0],[5,3]],[4,4]],[5,5]]`:

```
[1,1]
[2,2]
[3,3]
[4,4]
[5,5]
```

The final sum of this list is `[[[[5,0],[7,4]],[5,5]],[6,6]]`:

```
[1,1]
[2,2]
[3,3]
[4,4]
[5,5]
[6,6]
```

Here's a slightly larger example:

```
[[[0,[4,5]],[0,0]],[[[4,5],[2,6]],[9,5]]]
[7,[[[3,7],[4,3]],[[6,3],[8,8]]]]
[[2,[[0,8],[3,4]]],[[[6,7],1],[7,[1,6]]]]
[[[[2,4],7],[6,[0,5]]],[[[6,8],[2,8]],[[2,1],[4,5]]]]
[7,[5,[[3,8],[1,4]]]]
[[2,[2,2]],[8,[8,1]]]
[2,9]
[1,[[[9,3],9],[[9,0],[0,7]]]]
[[[5,[7,4]],7],1]
[[[[4,2],2],6],[8,7]]
```

The final sum `[[[[8,7],[7,7]],[[8,6],[7,7]]],[[[0,7],[6,6]],[8,7]]]` is found after adding up the above snailfish numbers:

```
  [[[0,[4,5]],[0,0]],[[[4,5],[2,6]],[9,5]]]
+ [7,[[[3,7],[4,3]],[[6,3],[8,8]]]]
= [[[[4,0],[5,4]],[[7,7],[6,0]]],[[8,[7,7]],[[7,9],[5,0]]]]

  [[[[4,0],[5,4]],[[7,7],[6,0]]],[[8,[7,7]],[[7,9],[5,0]]]]
+ [[2,[[0,8],[3,4]]],[[[6,7],1],[7,[1,6]]]]
= [[[[6,7],[6,7]],[[7,7],[0,7]]],[[[8,7],[7,7]],[[8,8],[8,0]]]]

  [[[[6,7],[6,7]],[[7,7],[0,7]]],[[[8,7],[7,7]],[[8,8],[8,0]]]]
+ [[[[2,4],7],[6,[0,5]]],[[[6,8],[2,8]],[[2,1],[4,5]]]]
= [[[[7,0],[7,7]],[[7,7],[7,8]]],[[[7,7],[8,8]],[[7,7],[8,7]]]]

  [[[[7,0],[7,7]],[[7,7],[7,8]]],[[[7,7],[8,8]],[[7,7],[8,7]]]]
+ [7,[5,[[3,8],[1,4]]]]
= [[[[7,7],[7,8]],[[9,5],[8,7]]],[[[6,8],[0,8]],[[9,9],[9,0]]]]

  [[[[7,7],[7,8]],[[9,5],[8,7]]],[[[6,8],[0,8]],[[9,9],[9,0]]]]
+ [[2,[2,2]],[8,[8,1]]]
= [[[[6,6],[6,6]],[[6,0],[6,7]]],[[[7,7],[8,9]],[8,[8,1]]]]

  [[[[6,6],[6,6]],[[6,0],[6,7]]],[[[7,7],[8,9]],[8,[8,1]]]]
+ [2,9]
= [[[[6,6],[7,7]],[[0,7],[7,7]]],[[[5,5],[5,6]],9]]

  [[[[6,6],[7,7]],[[0,7],[7,7]]],[[[5,5],[5,6]],9]]
+ [1,[[[9,3],9],[[9,0],[0,7]]]]
= [[[[7,8],[6,7]],[[6,8],[0,8]]],[[[7,7],[5,0]],[[5,5],[5,6]]]]

  [[[[7,8],[6,7]],[[6,8],[0,8]]],[[[7,7],[5,0]],[[5,5],[5,6]]]]
+ [[[5,[7,4]],7],1]
= [[[[7,7],[7,7]],[[8,7],[8,7]]],[[[7,0],[7,7]],9]]

  [[[[7,7],[7,7]],[[8,7],[8,7]]],[[[7,0],[7,7]],9]]
+ [[[[4,2],2],6],[8,7]]
= [[[[8,7],[7,7]],[[8,6],[7,7]]],[[[0,7],[6,6]],[8,7]]]
```

To check whether it's the right answer, the snailfish teacher only checks the magnitude of the final sum. The magnitude of a pair is 3 times the **magnitude** of its left element plus 2 times the magnitude of its right element. The magnitude of a regular number is just that number.

For example, the magnitude of `[9,1]` is `3*9 + 2*1 =`**`9`**; the magnitude of `[1,9]` is `3*1 + 2*9 =`**`21`**. Magnitude calculations are recursive: the magnitude of `[[9,1],[1,9]]` is `3*29 + 2*21 =`**`129`**.

Here are a few more magnitude examples:

- `[[1,2],[[3,4],5]]` becomes **`143`**.
- `[[[[0,7],4],[[7,8],[6,0]]],[8,1]]` becomes **`1384`**.
- `[[[[1,1],[2,2]],[3,3]],[4,4]]` becomes **`445`**.
- `[[[[3,0],[5,3]],[4,4]],[5,5]]` becomes **`791`**.
- `[[[[5,0],[7,4]],[5,5]],[6,6]]` becomes **`1137`**.
- `[[[[8,7],[7,7]],[[8,6],[7,7]]],[[[0,7],[6,6]],[8,7]]]` becomes **`3488`**.

So, given this example homework assignment:

```
[[[0,[5,8]],[[1,7],[9,6]]],[[4,[1,2]],[[1,4],2]]]
[[[5,[2,8]],4],[5,[[9,9],0]]]
[6,[[[6,2],[5,6]],[[7,6],[4,7]]]]
[[[6,[0,7]],[0,9]],[4,[9,[9,0]]]]
[[[7,[6,4]],[3,[1,3]]],[[[5,5],1],9]]
[[6,[[7,3],[3,2]]],[[[3,8],[5,7]],4]]
[[[[5,4],[7,7]],8],[[8,3],8]]
[[9,3],[[9,9],[6,[4,9]]]]
[[2,[[7,7],7]],[[5,8],[[9,3],[0,2]]]]
[[[[5,2],5],[8,[3,7]]],[[5,[7,5]],[4,4]]]
```

The final sum is:

```
[[[[6,6],[7,6]],[[7,7],[7,0]]],[[[7,7],[7,7]],[[7,8],[9,9]]]]
```

The magnitude of this final sum is **`4140`**.

Add up all of the snailfish numbers from the homework assignment in the order they appear. **What is the magnitude of the final sum?**

In [1]:
from json import loads

def load_input(filepath='./data/input_test.txt'):
    s_numbers = []
    with open(f'{filepath}', 'r') as input_f:
        for line in input_f.read().splitlines():
            s_numbers.append(loads(line))
            
    return s_numbers

load_input()

[[[[0, [5, 8]], [[1, 7], [9, 6]]], [[4, [1, 2]], [[1, 4], 2]]],
 [[[5, [2, 8]], 4], [5, [[9, 9], 0]]],
 [6, [[[6, 2], [5, 6]], [[7, 6], [4, 7]]]],
 [[[6, [0, 7]], [0, 9]], [4, [9, [9, 0]]]],
 [[[7, [6, 4]], [3, [1, 3]]], [[[5, 5], 1], 9]],
 [[6, [[7, 3], [3, 2]]], [[[3, 8], [5, 7]], 4]],
 [[[[5, 4], [7, 7]], 8], [[8, 3], 8]],
 [[9, 3], [[9, 9], [6, [4, 9]]]],
 [[2, [[7, 7], 7]], [[5, 8], [[9, 3], [0, 2]]]],
 [[[[5, 2], 5], [8, [3, 7]]], [[5, [7, 5]], [4, 4]]]]

In [2]:
def reduce(added_s_numbers):
    need_reduce = False
    
    new_s_numbers = []
    
    
    return new_s_numbers, need_reduce

def add_s_numbers(sn1, sn2):
    added_s_numbers = [sn1] + [sn2]
    
    return added_s_numbers

s_numbers = load_input()
add_s_numbers(s_numbers[0], s_numbers[1])

[[[[0, [5, 8]], [[1, 7], [9, 6]]], [[4, [1, 2]], [[1, 4], 2]]],
 [[[5, [2, 8]], 4], [5, [[9, 9], 0]]]]

In [3]:
class Node:

    def __init__(self, value=None, depth=0, parent=None):
        self.left = None
        self.right = None
        self.value = value
        self.parent = parent
        self.depth = depth
    
    
    @property
    def type(self):
        if isinstance(self.value, int):
            return 'int'
        else:
            return 'none'
    
    
    def print_tree(self):
        if self.left is not None:
            self.left.print_tree()
        print(' ' * 6 * self.depth + '->', self.value, '(' + str(self.depth) + ')')
        if self.right is not None:
            self.right.print_tree()
    
    
    def __repr__(self):
        return f'Node(value={self.value}, depth={self.depth})'

In [4]:
def build_tree(s_numbers, depth, parent=None):
    node = None
    if isinstance(s_numbers, int):
        node = Node(s_numbers, depth, parent)
    else:
        node = Node(None, depth, parent)
        node.left = build_tree(s_numbers[0], depth + 1, node)
        node.right = build_tree(s_numbers[1], depth + 1, node)
    
    return node
    
def detect_nodes_to_explode(node, node_list):
    if node.depth >= 4 and node.left is not None and \
        node.right is not None and node.left.type == 'int' and \
            node.right.type == 'int':
        node_list.append(node)
        return
    
    if node.left is not None:
        detect_nodes_to_explode(node.left, node_list)
    if node.right is not None:
        detect_nodes_to_explode(node.right, node_list)

def detect_nodes_to_split(node, node_list):
    if node.type == 'int' and node.value >= 10:
        node_list.append(node)
        return
    
    if node.left is not None:
        detect_nodes_to_split(node.left, node_list)
    if node.right is not None:
        detect_nodes_to_split(node.right, node_list)

def find_candidate(node, left_or_right):
    candidate = None
    
    def __find_candidate(node, left_or_right):
        if node.type == 'int':
            return node
        elif left_or_right == 'left':
            return __find_candidate(node.right, left_or_right)
        else:
            return __find_candidate(node.left, left_or_right)
    
    current_node = node
    node_parent = node.parent
    while current_node.parent is not None:
        if left_or_right == 'left':
            if node_parent.left == current_node:
                current_node = node_parent
                node_parent = node_parent.parent
            else:
                candidate = __find_candidate(node_parent.left, left_or_right)
                break
        else:
            if node_parent.right == current_node:
                current_node = node_parent
                node_parent = node_parent.parent
            else:
                candidate = __find_candidate(node_parent.right, left_or_right)
                break
    
    return candidate

def explode(node):
    candidate_l = find_candidate(node, 'left')
    candidate_r = find_candidate(node, 'right')
    # print(f'candidate_l: {candidate_l}, candidate_r: {candidate_r}')
    if candidate_l is not None:
        candidate_l.value += node.left.value
    if candidate_r is not None:
        candidate_r.value += node.right.value
    del node.left
    del node.right
    node.left = None
    node.right = None
    node.value = 0
    
    return candidate_l, candidate_r
    

def split(node):
    new_node_l = Node(int(node.value / 2), node.depth + 1, node)
    new_node_r = Node(int(node.value / 2) + node.value % 2, node.depth + 1, node)
    node.value = None
    node.left = new_node_l
    node.right = new_node_r

def explode_and_split(node):
    node_list_explode = []
    node_list_split = []
    detect_nodes_to_explode(node, node_list_explode)
    if node_list_explode:
        candidate_l, candidate_r = explode(node_list_explode[0])
        
        return True

    node_list_split = []
    detect_nodes_to_split(node, node_list_split)
    if node_list_split:
        split(node_list_split[0])
        
        return True
    
    return False

In [5]:
s_numbers = [[[[[1,1],[2,2]],[3,3]],[4,4]],[5,5]]
root = build_tree(s_numbers, 0)
# root.print_tree()

while True:
    # root.print_tree()
    has_exploded_or_split = explode_and_split(root)
    
    if not has_exploded_or_split:
        break
        
root.print_tree()

                        -> 3 (4)
                  -> None (3)
                        -> 0 (4)
            -> None (2)
                        -> 5 (4)
                  -> None (3)
                        -> 3 (4)
      -> None (1)
                  -> 4 (3)
            -> None (2)
                  -> 4 (3)
-> None (0)
            -> 5 (2)
      -> None (1)
            -> 5 (2)


In [6]:
def add_s_numbers(s_num_1, s_num_2):
    root = Node()
    root.left = s_num_1
    root.right = s_num_2
    root.left.parent = root.right.parent = root
    
    def update_tree_depth(node):
        node.depth += 1
        
        if node.left is not None:
            update_tree_depth(node.left)
        if node.right is not None:
            update_tree_depth(node.right)
            
    update_tree_depth(root.left)
    update_tree_depth(root.right)
    
    return root
    
s_numbers_1 = build_tree([1,2], 0)
s_numbers_2 = build_tree([[1,2],3], 0)
added_s_numbers = add_s_numbers(s_numbers_1, s_numbers_2)
added_s_numbers.print_tree()

            -> 1 (2)
      -> None (1)
            -> 2 (2)
-> None (0)
                  -> 1 (3)
            -> None (2)
                  -> 2 (3)
      -> None (1)
            -> 3 (2)


In [7]:
s_numbers = load_input()

added_s_numbers = build_tree(s_numbers[0], 0)

for i in range(1, len(s_numbers)):
    added_s_numbers = add_s_numbers(added_s_numbers, build_tree(s_numbers[i], 0))
    
    while True:
        has_exploded_or_split = explode_and_split(added_s_numbers)

        if not has_exploded_or_split:
            break

added_s_numbers.print_tree()

                        -> 6 (4)
                  -> None (3)
                        -> 6 (4)
            -> None (2)
                        -> 7 (4)
                  -> None (3)
                        -> 6 (4)
      -> None (1)
                        -> 7 (4)
                  -> None (3)
                        -> 7 (4)
            -> None (2)
                        -> 7 (4)
                  -> None (3)
                        -> 0 (4)
-> None (0)
                        -> 7 (4)
                  -> None (3)
                        -> 7 (4)
            -> None (2)
                        -> 7 (4)
                  -> None (3)
                        -> 7 (4)
      -> None (1)
                        -> 7 (4)
                  -> None (3)
                        -> 8 (4)
            -> None (2)
                        -> 9 (4)
                  -> None (3)
                        -> 9 (4)


In [8]:
def magnitude(root: Node):
    def calc_magnitude(node: Node):
        return 2 * node.right.value + 3 * node.left.value
    
    def process_magnitude(node: Node):
        if node.left is not None and \
            node.right is not None and node.left.type == 'int' and \
                node.right.type == 'int':
            node.value = calc_magnitude(node)
            node.left.parent = node.right.parent = None
            node.left = node.right = None
        else:
            if node.left is not None:
                process_magnitude(node.left)
            if node.right is not None:
                process_magnitude(node.right)

    while root.type != 'int':
        process_magnitude(root)

    return root.value

print(magnitude(build_tree([[1,2],[[3,4],5]], 0)))
magnitude(added_s_numbers)

143


4140

In [9]:
s_numbers = load_input('./data/input.txt')

added_s_numbers = build_tree(s_numbers[0], 0)

for i in range(1, len(s_numbers)):
    added_s_numbers = add_s_numbers(added_s_numbers, build_tree(s_numbers[i], 0))
    
    while True:
        has_exploded_or_split = explode_and_split(added_s_numbers)

        if not has_exploded_or_split:
            break

added_s_numbers.print_tree()
magnitude(added_s_numbers)

                        -> 7 (4)
                  -> None (3)
                        -> 6 (4)
            -> None (2)
                        -> 0 (4)
                  -> None (3)
                        -> 7 (4)
      -> None (1)
                        -> 6 (4)
                  -> None (3)
                        -> 7 (4)
            -> None (2)
                        -> 8 (4)
                  -> None (3)
                        -> 9 (4)
-> None (0)
                        -> 4 (4)
                  -> None (3)
                        -> 0 (4)
            -> None (2)
                  -> 4 (3)
      -> None (1)
                        -> 7 (4)
                  -> None (3)
                        -> 9 (4)
            -> None (2)
                        -> 4 (4)
                  -> None (3)
                        -> 0 (4)


3051

Your puzzle answer was `3051`.

## Part two

You notice a second question on the back of the homework assignment:

What is the largest magnitude you can get from adding only two of the snailfish numbers?

Note that snailfish addition is not [commutative](https://en.wikipedia.org/wiki/Commutative_property) - that is, `x + y` and `y + x` can produce different results.

Again considering the last example homework assignment above:

```
[[[0,[5,8]],[[1,7],[9,6]]],[[4,[1,2]],[[1,4],2]]]
[[[5,[2,8]],4],[5,[[9,9],0]]]
[6,[[[6,2],[5,6]],[[7,6],[4,7]]]]
[[[6,[0,7]],[0,9]],[4,[9,[9,0]]]]
[[[7,[6,4]],[3,[1,3]]],[[[5,5],1],9]]
[[6,[[7,3],[3,2]]],[[[3,8],[5,7]],4]]
[[[[5,4],[7,7]],8],[[8,3],8]]
[[9,3],[[9,9],[6,[4,9]]]]
[[2,[[7,7],7]],[[5,8],[[9,3],[0,2]]]]
[[[[5,2],5],[8,[3,7]]],[[5,[7,5]],[4,4]]]
```

The largest magnitude of the sum of any two snailfish numbers in this list is **`3993`**. This is the magnitude of `[[2,[[7,7],7]],[[5,8],[[9,3],[0,2]]]] + [[[0,[5,8]],[[1,7],[9,6]]],[[4,[1,2]],[[1,4],2]]]`, which reduces to `[[[[7,8],[6,6]],[[6,0],[7,7]]],[[[7,8],[8,8]],[[7,9],[0,6]]]]`.

**What is the largest magnitude of any sum of two different snailfish numbers from the homework assignment?**

In [10]:
s_numbers = load_input()

magnitudes = set()

for i in range(0, len(s_numbers) - 1):
    for j in range(i, len(s_numbers)):
        added_s_numbers = add_s_numbers(build_tree(s_numbers[i], 0), build_tree(s_numbers[j], 0))
        while True:
            has_exploded_or_split = explode_and_split(added_s_numbers)

            if not has_exploded_or_split:
                break
        magnitudes.add(magnitude(added_s_numbers))

        added_s_numbers = add_s_numbers(build_tree(s_numbers[j], 0), build_tree(s_numbers[i], 0))
        while True:
            has_exploded_or_split = explode_and_split(added_s_numbers)

            if not has_exploded_or_split:
                break
        magnitudes.add(magnitude(added_s_numbers))

max(magnitudes)

3993

In [11]:
s_numbers = load_input('./data/input.txt')

magnitudes = set()

for i in range(0, len(s_numbers) - 1):
    for j in range(i, len(s_numbers)):
        added_s_numbers = add_s_numbers(build_tree(s_numbers[i], 0), build_tree(s_numbers[j], 0))
        while True:
            has_exploded_or_split = explode_and_split(added_s_numbers)

            if not has_exploded_or_split:
                break
        magnitudes.add(magnitude(added_s_numbers))

        added_s_numbers = add_s_numbers(build_tree(s_numbers[j], 0), build_tree(s_numbers[i], 0))
        while True:
            has_exploded_or_split = explode_and_split(added_s_numbers)

            if not has_exploded_or_split:
                break
        magnitudes.add(magnitude(added_s_numbers))

max(magnitudes)

4812

Your puzzle answer was `4812`.